fix #6241: "mappings to <Nop> wait forever"

hardenedapple opened the pull request:
vgetorpeek() uses `timedout` to signify whether it should wait for more
keys to complete a partial mapping or not.
Once this has been set, the function expands whatever is the longest
full mapping in the typebuf without waiting for more characters from the
user.
If this expansion results in another possibly incomplete mapping, then
the 'timedout' variable is left-over, and ignores these possibilities.

Bug #6241 was caused because <Nop> is expanded as an empty string into
the typbuffer. This means that vgetorpeek() doesn't return and keeps
waiting for a useful character to come.
Hence vim will take the next key without waiting to see if it is part of
a multi-key map if the user doesn't type quick enough to put more than
one character into the typebuffer at once.
Because in that bug there is a single-key map for <Nop>, this is the one
that is used, perpetuating the problem.

timedout can be reset between checking the typebuffer for characters and
getting input from the user because:
timedout is only used when looking in the typebuffer.
The only time that it could possibly matter that timedout is reset after
looking in the typebuffer is if we don't find a character in the
typbuffer as this is the only time we don't either exit the for loop or
`continue` on the next iteration.
timedout must have previously been set to TRUE, otherwise resetting it
to FALSE doesn't matter.
This means there must be no characters in the typebuffer.

When attempting to get a key from the user, the second chunk of
vgetorpeek() must not `break` out of the for loop with a condition such
that the while() loop will not continue on its iteration.
Otherwise timedout won't be read again (it's only read once, while
checking for mappings).
There are a few places in this chunk that `continue` is used after
setting timedout to TRUE. These won't be affected by the reset of
timedout.
The only times that the loop can continue without timedout being reset,
are either when the end of an input script being reached, or some bytes
are successfully read from the user.

Using these limits, the only time that resetting timedout changes the
behaviour of the function is when a mapping was not completed because of
timeout, and a smaller mapping that was a substring of the timed out
mapping was taken, and (whether after multiple map expansions or just
the one) resulted in a <Nop>.
This had to be a mapping so that the `break` clause is not invoked.
If the mapping actually inserted something, then vgetorpeek() would look
again in typebuf until it either found an unmapped key or couldn't find
anything. If it found an unmapped key, then this key would be returned,
and a reset of timedout between getting a character from typebuf and
from the user doesn't matter.
The only time it can't find anything is when something is mapped to
<Nop>.

In this case the control goes into the section getting input from the
user. Here, there are two possible methods that we could get back into
look_in_typebuf() without resetting timedout.
Either we were reading from a scriptfile, and that now closes, or we
get new input from the user.

When getting new input from the user, we want timedout to be reset, because
otherwise this would mean the user can't use any more maps.

When finishing getting input from the scriptfile I think the same
applies: we want timedout to be reset so that the user (or the next
scriptfile) doesn't get state carried over from the previous file.

hardenedapple committed:
close #6241

vgetorpeek() uses `timedout` to signify whether it should wait for more
keys to complete a partial mapping or not.
Once this has been set, the function expands whatever is the longest
full mapping in the typebuf without waiting for more characters from the
user.
If this expansion results in another possibly incomplete mapping, then
the 'timedout' variable is left-over, and ignores these possibilities.

Bug #6241 was caused because <Nop> is expanded as an empty string into
the typbuffer. This means that vgetorpeek() doesn't return and keeps
waiting for a useful character to come.
Hence vim will take the next key without waiting to see if it is part of
a multi-key map if the user doesn't type quick enough to put more than
one character into the typebuffer at once.
Because in that bug there is a single-key map for <Nop>, this is the one
that is used, perpetuating the problem.

timedout can be reset between checking the typebuffer for characters and
getting input from the user because:
timedout is only used when looking in the typebuffer.
The only time that it could possibly matter that timedout is reset after
looking in the typebuffer is if we don't find a character in the
typbuffer as this is the only time we don't either exit the for loop or
`continue` on the next iteration.
timedout must have previously been set to TRUE, otherwise resetting it
to FALSE doesn't matter.
This means there must be no characters in the typebuffer.

When attempting to get a key from the user, the second chunk of
vgetorpeek() must not `break` out of the for loop with a condition such
that the while() loop will not continue on its iteration.
Otherwise timedout won't be read again (it's only read once, while
checking for mappings).
There are a few places in this chunk that `continue` is used after
setting timedout to TRUE. These won't be affected by the reset of
timedout.
The only times that the loop can continue without timedout being reset,
are either when the end of an input script being reached, or some bytes
are successfully read from the user.

Using these limits, the only time that resetting timedout changes the
behaviour of the function is when a mapping was not completed because of
timeout, and a smaller mapping that was a substring of the timed out
mapping was taken, and (whether after multiple map expansions or just
the one) resulted in a <Nop>.
This had to be a mapping so that the `break` clause is not invoked.
If the mapping actually inserted something, then vgetorpeek() would look
again in typebuf until it either found an unmapped key or couldn't find
anything. If it found an unmapped key, then this key would be returned,
and a reset of timedout between getting a character from typebuf and
from the user doesn't matter.
The only time it can't find anything is when something is mapped to
<Nop>.

In this case the control goes into the section getting input from the
user. Here, there are two possible methods that we could get back into
look_in_typebuf() without resetting timedout.
Either we were reading from a scriptfile, and that now closes, or we
get new input from the user.

When getting new input from the user, we want timedout to be reset, because
otherwise this would mean the user can't use any more maps.

When finishing getting input from the scriptfile I think the same
applies: we want timedout to be reset so that the user (or the next
scriptfile) doesn't get state carried over from the previous file.

justinmk commented:
Do you plan to add a test?

hardenedapple commented:
I was just writing up my thoughts on testing -- I don't know how to write a _nice_ test for this.

I need to send some keys, wait for vim to get into `vgetorpeek()`, read the shortened map, fail to get any more input from the user, and expand a map into `<Nop>`.
*Then* (and not before) I need to send some more keys to the vim instance.

The best idea I have at the moment is just to wait a while with some pause in lua, I'm looking into it, but I don't like it.

justinmk commented:
It's ok if it's not nice. How about setting `timeout=1`, send a key mapped to `<Nop>`, then use `helpers.retry()` to test the result of `nvim_get_mode()` (#6247) ?

hardenedapple commented:
I don't know whether #6247 would help (because I don't understand it).

I guess it wouldn't, because the vim process can't leave `vgetorpeek()` between entering before the `<Nop>` expansion and the next set of keys added. So I guess it can't handle the `nvim_get_mode()` api call (unless maybe `os_inchar()` does something?).

hardenedapple commented:
Also, I thought I'd mention:

This particular bug is essentially a special case of the following behaviour:
```vimL
nnoremap aaaa :echom 'aaaa mapping'<CR>
nnoremap bb :echom 'just to ensure timeout matters'<CR>
nmap b aaa
```
If you press `b`, and wait for the timeout, then it inserts `aa` (i.e. runs `aaa`) without waiting to see if you're going to type `a` to finish the new mapping.

I've made sure this behaviour stays around, but thought I'd ask just in case people would prefer vim waited for another timeout to see if the `aaa` inserted by the `b` mapping is finished by the user with another `a`.

justinmk commented:
> I've made sure this behaviour stays around, but thought I'd ask just in case people would prefer vim waited for another timeout to see if the aaa inserted by the b mapping is finished by the user with another a.

I think that would need to be decided upstream. cc @brammool @chrisbra

hardenedapple committed:
Add tests for <Nop> timeout

Not great (does rely on a timeout in one place).
Requires `mkfifo` and `timeout`.

hardenedapple commented:
I've added tests, but they're ugly. They rely on timing between the tests process and the vim process, so I've made the time windows very large to decrease the probability of failures.

Any better ideas are most certainly welcome.

justinmk commented on test/functional/normal/map_spec.lua:
```diff
@@ -0,0 +1,49 @@
+    -- before vgetorpeek() has timed out.
+    -- Because we're using feed(), and it puts characters into the typebuf all
+    -- at once, we have no problem triggering the longer mappings.
+    execute('set timeout', 'set timeoutlen=1000')
```
why not set it to 1ms, so that the sleeps don't need to be so long?

brammool commented:
I can add a test, but measuring timeouts can't be done reliably.

func Test_map_timeout()
  nnoremap aaaa :let got_aaaa = 1<CR>
  nnoremap bb :let got_bb = 1<CR>
  nmap b aaa
  new
  func ExitInsert(timer)
    let g:line = getline(1)
    call feedkeys("\<Esc>", "t")
  endfunc
  set timeout timeoutlen=200
  call timer_start(300, 'ExitInsert')
  " After the 'b' Vim waits for another character to see if it matches 'bb'.
  " When it times out it is expanded to "aaa", but there is no wait for
  " "aaaa".  Can't check that reliably though.
  call feedkeys("b", "xt!")
  call assert_equal("aa", g:line)
  call assert_false(exists('got_aaa'))
  call assert_false(exists('got_bb'))

  bwipe!
  nunmap aaaa
  nunmap bb
  nunmap b
  set timeoutlen&
  delfunc ExitInsert
endfunc

hardenedapple commented:
@brammool Yeah, that's the same problem I'm having ... 

I take it you think the behaviour of not waiting for the user to finish the possible `aaaa` mapping is to be kept?

brammool commented:
I have not heard of a reason why changing the behavior would be better in practice. General rule: don't fix things that aren't broken.

justinmk commented:
@hardenedapple is https://github.com/vim/vim/commit/b7637c44c26b057d1f3721d932bbab06d9f74393 related to this, perhaps it can be used as the test?

hardenedapple commented:
That test is testing for the same thing as my `does not wait for timeout a second time` test.

It relies on timing (hence mentioning `Can't check that reliably though`), while my test doesn't.

In my version I `set notimeout` in the `<expr>` map, so I don't have to worry about the second map being expanded before I send the second key, and I have communication between the test and the map expansion through a fifo so I don't have to worry about sending the `a` key before the `b` mapping has been applied due to a timeout.

justinmk commented:
@hardenedapple I don't see `<nop>` in the tests. Should it be?

hardenedapple commented:
If I remember correctly, I used the ExprMap function with an empty string
to simulate a nop.
I needed the expression map to a function rather than a simple nop map so i
could manipulate timeouts while vgetorpeek was running.

On Tue, 1 May 2018, 10:05 Justin M. Keyes, <notifications@github.com> wrote:

> @hardenedapple <https://github.com/hardenedapple> I don't see <nop> in
> the tests. Should it be?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/6419#issuecomment-385629849>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD4c3H3woqKLwotGIEfTu8SQTWdS1sWgks5tuCVXgaJpZM4MwdjA>
> .
>

janlazo commented:
Requires a rebase or something else to test the changes via Github Actions.

janlazo commented:
Reopen if https://github.com/neovim/neovim/commit/3b744f1ea273949069523a7d66d43f9ac1115317 is insufficient.
